### PR TITLE
[Train] Clarify shuffle documentation in `prepare_data_loader`

### LINKED ISSUE #22406: [Bug] Cannot disable shuffling with `prepare_data_loader`
### Search before asking

- [X] I searched the [issues](https://github.com/ray-project/ray/issues) and found no similar issues.

### Ray Component

Ray Train

### What happened + What you expected to happen

I want to prepare my data loader for distributed inference, but I don't want to shuffle the samples (as is typical for model evaluation).

```
test_datatset = torchvision.datasets.ImageFolder("~/data/TinyImageNet/val")
test_data_loader = DataLoader(test_dataset, shuffle=False)
test_data_loader = prepare_data_loader(test_data_loader)  # This data loader will shuffle indices!
```

However, `prepare_data_loader` will always shuffle the indices unless I pass in a `SequentialSampler` to `prepare_data_loader`. The relevant code snippet from the source code is included below.

```
def with_sampler(loader):
        ...

        # Always evaluates to `True` unless a `SequentialSampler` is passed in!
        shuffle = not isinstance(loader.sampler, SequentialSampler)

        data_loader_args = {
            ...,
            "sampler": DistributedSampler(loader.dataset, shuffle=shuffle),
        }
        return DataLoader(**data_loader_args)

data_loader = with_sampler(data_loader)
```

My current work around is to manually create and pass in a `DistributedSampler`.

### Versions / Dependencies

Python: 3.8
OS: Ubuntu
Ray: 1.10

### Reproduction script

TK

### Anything else

_No response_

### Are you willing to submit a PR?

- [X] Yes I am willing to submit a PR!

### PR DESCRIPTION
**Pull request by bveeramani**
## Why are these changes needed?

We essentially use a hack to determine whether shuffling should be enabled in `prepare_data_loader`. I've clarified the documentation so the hack is easier to understand.

## Related issue number

Closes #22406 

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [x] This PR is not tested :(

### CODE REVIEW

File: python/ray/train/torch.py [hunk 1]
```diff
@@ -316,14 +316,16 @@ def prepare_data_loader(
...
-            shuffle = not isinstance(loader.sampler, SequentialSampler)
+            # If you're using a sampler, the DataLoader shuffle flag must be set to
+            # False. Shuffling is instead determined by the shuffle argument passed to
+            # the DistributedSampler constructor.
+
+            # If no sampler is passed to the DataLoader constructor, Torch constructs a
+            # default sampler. The default sampler is a RandomSampler if shuffling is
+            # enabled and a SequentialSampler otherwise. There is no DataLoader.shuffle
+            # attribute. So, we identify whether shuffling is enabled by checking the
+            # default sampler type.
```
Comment (matthewdeng) [inline python/ray/train/torch.py:327]:
```suggestion
            # enabled and a SequentialSampler otherwise. DataLoader does not have a
            # shuffle attribute, so we instead identify whether shuffling is enabled by 
            # checking the default sampler type.
```

File: python/ray/train/torch.py [hunk 2]
```diff
@@ -316,14 +316,16 @@ def prepare_data_loader(
...
+            shuffle = isinstance(loader.sampler, RandomSampler)
```
Comment (matthewdeng) [inline python/ray/train/torch.py:328]: Can we log a message when the Sampler is neither `RandomSampler` nor `SequentialSampler`, and indicate to the user what the default behavior is?

Comment (bveeramani) [reply python/ray/train/torch.py:328]: What do you mean by indicating the default behavior to the user?

It should always be one of the two, so we can probably `assert isinstance(loader.sampler, (RandomSampler, SequentialSampler))`

Comment (matthewdeng) [reply python/ray/train/torch.py:328]: Ah but if the original `DataLoader` can be initialized with a `sampler` instead of the `shuffle` flag, right?

Comment (amogkam) [reply python/ray/train/torch.py:328]: Yep lets log this, saying a sampler has been already set, but overriding with DistributedSampler. You can disable this by setting `with_sampler` to False in `prepare_data_loader`.

Comment (amogkam) [reply python/ray/train/torch.py:328]: Also, I believe the reason we had `shuffle = not isinstance(loader.sampler, SequentialSampler)` before is because there are multiple types of random samplers (for example `WeightedRandomSampler`), so we just treat all those as `shuffle=True` case.

File: python/ray/train/torch.py [hunk 3]
```diff
@@ -316,15 +317,26 @@ def prepare_data_loader(
...
+            # checking the default sampler type.
             shuffle = not isinstance(loader.sampler, SequentialSampler)
 
+            using_default_sampler = isinstance(
+                loader.sampler, (SequentialSampler, RandomSampler)
+            )
+            if not using_default_sampler:
+                logger.warn(
+                    f"The {loader.sampler.__class__.__name__} will be overwritten with a DistributedSampler."
+                    f"You can disable this by setting with_sampler to False in prepare_data_loader"
```
Comment (amogkam) [inline python/ray/train/torch.py:337]:
```suggestion
                    f"You can disable this by setting `with_sampler` to False in `prepare_data_loader`"
```

File: python/ray/train/torch.py [hunk 4]
```diff
@@ -316,15 +317,26 @@ def prepare_data_loader(
...
+            # enabled and a SequentialSampler otherwise. DataLoader does not have a
+            # shuffle attribute, so we instead identify whether shuffling is enabled by
+            # checking the default sampler type.
             shuffle = not isinstance(loader.sampler, SequentialSampler)
 
+            using_default_sampler = isinstance(
+                loader.sampler, (SequentialSampler, RandomSampler)
+            )
+            if not using_default_sampler:
+                logger.warn(
```
Comment (amogkam) [inline python/ray/train/torch.py:335]: Should we only log if `train.world_rank()==0`?

**Review by amogkam — APPROVED** [review]
Nice! Couple minor comments.

**Review by matthewdeng — APPROVED** [review]

**Comment by amogkam** [comment]
@bveeramani could you resolve the merge conflicts here, and then we can merge this in?
